src/clusterfuzz/_internal/system/archive.py

```diff
@@ -407,6 +412,103 @@ class ArchiveType:
   TAR_LZMA = 3
 
 
+@dataclasses.dataclass
+class CacheBlock:
+  """Represents a cache entry for the HttpZipFile."""
+  start: int
+  end: int
+  content: Sequence[bytes]
+
+
+class HttpZipFile:
+  """This class is a very simple file-object representation of a file over HTTP.
+  It uses the 'Accept-Ranges' feature of HTTP to fetch parts (or all) of the
+  file.
+  """
+
+  @staticmethod
+  def is_uri_compatible(uri: str) -> bool:
+    try:
+      res = requests.head(uri, timeout=5)
+      return res.headers.get('Accept-Ranges') is not None
+    except:
+      return False
+
+  def __init__(self, uri):
+    self.uri = uri
+    self.session = requests.Session()
+    # This slows down the overall performances.
+    self.session.headers.pop('Accept-Encoding')
+    resp = self.session.head(self.uri)
+    self.file_size = int(resp.headers.get('Content-Length', default=0))
+    self._current_block = CacheBlock(0, 0, [])
+    self._pos = 0
+    assert resp.headers.get('Accept-Ranges') is not None
+
+  def seekable(self) -> bool:
+    """Whether this is seekable.
+    """
+    return True
+
+  def seek(self, offset: int, from_what: int = 0):
+    """Provides a seek implementation.
+
+    Args:
+        offset: the offset
+        from_what: from where the offset should be computed. Defaults to 0.
+    """
+    if from_what == 0:
+      self._pos = offset
+    elif from_what == 1:
+      self._pos = self._pos + offset
+    else:
+      self._pos = self.file_size + offset
+    if self._pos > self.file_size:
+      self._pos = self.file_size
+    self._pos = max(self._pos, 0)
+
+  def tell(self) -> int:
+    """Provides a tell implementation. Returns the current curso position.
+
+    Returns:
+        the current cursor position.
+    """
+    return self._pos
+
+  def _fetch_from_http(self, start, end) -> Sequence[bytes]:
+    resp = self.session.get(self.uri, headers={'Range': f'bytes={start}-{end}'})
+    return resp.content
+
+  def _fetch_from_cache(self, start, end) -> Sequence[bytes]:
+    if self._current_block.start > start or self._current_block.end < end:
+      read_ahead_end = min(self.file_size - 1, end + REMOTE_HTTP_MIN_READ_SIZE)
+      self._current_block = CacheBlock(
+          start, read_ahead_end, self._fetch_from_http(start, read_ahead_end))
+    inner_start = start - self._current_block.start
+    inner_end = end - self._current_block.start
+    return self._current_block.content[inner_start:inner_end + 1]
+
+  def read(self, size=-1) -> Sequence[bytes]:
+    """Read into this file-object.
+
+    Args:
+        size: the size of the read. If not specified, reads all.
+
+    Returns:
+        a sequence of bytes.
+    """
+    if size == -1:
+      size = self.file_size - self._pos
+    read_size = min(self.file_size - self._pos, size)
+    end_range = self._pos + read_size - 1
+    if read_size > REMOTE_HTTP_MIN_READ_SIZE:
+      content = self._fetch_from_http(self._pos, end_range)
```

Comment:
This means we won't cache these bytes - is that intentional? Why not always delegate to `_fetch_from_cache()` and let it determine whether or not to fetch more bytes from the wire?

Reply:
Yep, that's because the way those archive library works is that they only read what's needed. If the read is greater than 50 MB, that means we're reading a huge file, and we won't need it in following reads. This is only useful for small reads tbh.

Reply:
Ah, I see. Can you explain that in a comment for the future reader? 

Optional comment / musing follows below:

It results in a behavior I find counterintuitive, though that does not terribly matter:

* read(49M) -> fetches 99MB and caches it
* read(50M) -> fetches 50MB, does not cache it

I think this could be addressed by doing something like:

```py
HTTP_MAX_BYTES = 100 * 1024 * 1024

def fetch_size(size):
  """How many bytes to fetch over HTTP to serve `read(size)`."""
  cache_size = min(size + HTTP_READAHEAD_BYTES, HTTP_MAX_BYTES)
  return max(size, cache_size)
```

Then at least you don't get a cliff effect for network fetches. But is it useful? Probably not?

Reply:
Thanks for adding a comment. It only explains what is happening, but not *why*, which is really what's interesting here. In other words, I can see for myself from reading the code that we don't cache large requests, but it does not tell me why that is. You explained it well in your PR comment, can you do that in the code?